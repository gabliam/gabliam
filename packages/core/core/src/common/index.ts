export * from './utils';
export * from './interfaces';
