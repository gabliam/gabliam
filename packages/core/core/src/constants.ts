/** Type in registry */
export const TYPE = {
  Service: 'ServiceType',
  Config: 'ConfigType',
  PreDestroy: 'PreDestroy',
};

export const METADATA_KEY = {
  bean: '_bean',
  onMissingBean: '_onMissingBean',
  init: '_init',
  beforeCreate: '_beforeCreate',
  value: '_value',
  register: '_register',
  scan: '_scan',
  config: '_config',
  service: '_service',
  plugin: '_plugin',
  injectContainer: '_inject_container',
  preDestroy: '_preDestroy',
};

/**
 * Inject container key in descriptor
 */
export const INJECT_CONTAINER_KEY = '$$container';

export const ORDER_CONFIG = {
  Core: 50,
  Plugin: 100,
  Config: 150,
};

export const ERRORS_MSGS = {
  DUPLICATED_DECORATOR: `Cannot apply decorator multiple times.`,
  DUPLICATED_CONFIG_DECORATOR: `Cannot apply @Config, @PluginConfig or @CorePlugin decorator multiple times.`,
  DUPLICATED_PLUGIN_DECORATOR: `Cannot apply @Plugin decorator multiple times.`,
  DUPLICATED_APPLICATION_DECORATOR: `Cannot apply @Application decorator multiple times.`,
};

export const APP_CONFIG = Symbol('APP_CONFIG');

export const CORE_CONFIG = Symbol('CORE_CONFIG');

export const VALUE_EXTRACTOR = Symbol('ValueExtractor');
