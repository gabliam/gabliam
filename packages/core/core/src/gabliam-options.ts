import { GabliamConfig } from './interfaces';
import * as _ from 'lodash';

const DEFAULT_CONFIG: GabliamConfig = {
  config: process.env.GABLIAM_CONFIG_PATH,
};

export const getGabliamConfig = (
  options?: Partial<GabliamConfig> | string
): GabliamConfig => {
  if (options === undefined) {
    return DEFAULT_CONFIG;
  } else {
    if (_.isString(options)) {
      return {
        ...DEFAULT_CONFIG,
        config: options,
      };
    } else {
      return {
        ...DEFAULT_CONFIG,
        ...options,
      };
    }
  }
};
