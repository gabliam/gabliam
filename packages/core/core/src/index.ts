import 'reflect-metadata';

export {
  decorate,
  inject,
  injectable,
  interfaces as inversifyInterfaces,
  multiInject,
  named,
  optional,
  postConstruct,
  tagged,
  targetName,
  unmanaged,
} from 'inversify';
export * from './common';
export {
  APP_CONFIG,
  CORE_CONFIG,
  INJECT_CONTAINER_KEY,
  VALUE_EXTRACTOR,
} from './constants';
export * from './container';
export * from './decorator';
export * from './errors';
export * from './gabliam';
export * from './gabliam-options';
export * from './gabliam-utils';
export * from './interfaces';
export * from './joi';
export { configResolver, FileLoader, Loader, Resolver } from './loaders';
export * from './metadatas';
export * from './reflection';
export * from './registry';
export * from './value-extractor';
