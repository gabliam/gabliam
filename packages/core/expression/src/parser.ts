import { generate } from 'escodegen';
import {
  Expression,
  UnaryExpression,
  ArrayExpression,
  SpreadElement,
  ObjectExpression,
  Pattern,
  BinaryExpression,
  LogicalExpression,
  Identifier,
  ThisExpression,
  CallExpression,
  Super,
  MemberExpression,
  ConditionalExpression,
  ExpressionStatement,
  ReturnStatement,
  FunctionExpression,
  TemplateLiteral,
  TaggedTemplateExpression,
  TemplateElement,
  Statement,
} from 'estree';

const FAIL = Symbol('FAIL');

export const IS_STRING = Symbol('IS_STRING');

export class Parser {
  constructor(private ast: Expression) {}

  parse<T = any>(vars: object = {}): T | undefined | null {
    const result = this.walk(this.ast, vars);
    return result === FAIL ? undefined : result;
  }

  private parseUnary(node: UnaryExpression, vars: object) {
    const val = this.walk(node.argument, vars);
    switch (node.operator) {
      case '+':
        return +val;
      case '-':
        return -val;
      case '~':
        return ~val;
      case '!':
        return !val;
      case 'typeof':
        return typeof val;
      case 'void':
      case 'delete':
        return undefined;
      default:
        return FAIL;
    }
  }

  private parseArray(node: ArrayExpression, vars: object) {
    const xs: any[] = [];
    for (const nodeElement of node.elements) {
      const x = this.walk(nodeElement, vars);
      if (x === FAIL) {
        return FAIL;
      }
      xs.push(x);
    }
    return xs;
  }

  private parseObject(node: ObjectExpression, vars: object) {
    const obj = {};
    for (let i = 0; i < node.properties.length; i++) {
      const prop: any = node.properties[i];
      const value: any =
        prop.value === null ? prop.value : this.walk(prop.value, vars);
      if (value === FAIL) {
        return FAIL;
      }
      const key: any = prop.key === null ? prop.key : this.walk(prop.key, vars);
      if (key === FAIL) {
        return FAIL;
      }
      (<any>obj)[key] = value;
    }
    return obj;
  }

  private parseLeftRight(
    node: BinaryExpression | LogicalExpression,
    vars: object,
  ) {
    const l = this.walk(node.left, vars);
    if (l === FAIL) {
      return [FAIL];
    }
    const r = this.walk(node.right, vars);
    if (r === FAIL) {
      return [FAIL];
    }
    return [l, r];
  }

  private parseBinary(node: BinaryExpression, vars: object) {
    const [l, r] = this.parseLeftRight(node, vars);
    if (l === FAIL) {
      return FAIL;
    }
    if (
      null === l &&
      undefined === r &&
      (<any>node.left).raw &&
      node.right.type === 'Identifier'
    ) {
      return IS_STRING;
    }

    switch (node.operator) {
      case '==':
        return l == r;

      case '!=':
        return l != r;

      case '===':
        return l === r;

      case '!==':
        return l !== r;

      case '<':
        return l < r;

      case '<=':
        return l <= r;

      case '>':
        return l > r;

      case '>=':
        return l >= r;

      case '<<':
        return l << r;

      case '>>':
        return l >> r;

      case '>>>':
        return l >>> r;

      case '+':
        return l + r;

      case '-':
        return l - r;

      case '*':
        return l * r;

      case '/':
        return l / r;

      case '%':
        return l % r;

      case '**':
        return l ** r;

      case '|':
        return l | r;

      case '^':
        return l ^ r;

      case '&':
        return l & r;

      case 'in':
        return l in r;

      case 'instanceof':
        return l instanceof r;
    }
  }

  private parseLogical(node: LogicalExpression, vars: object) {
    const l = this.walk(node.left, vars);
    if (l === FAIL) {
      return FAIL;
    }

    if (l === false && node.operator === '&&') {
      return false;
    }

    const r = this.walk(node.right, vars);

    switch (node.operator) {
      case '||':
        return l || r;
      case '&&':
        return l && r;
    }
  }

  private parseIdentifier(node: Identifier, vars: object) {
    if ({}.hasOwnProperty.call(vars, node.name)) {
      return (<any>vars)[node.name];
    }
    if (global.hasOwnProperty(node.name)) {
      return (<any>global)[node.name];
    }
    return undefined;
  }

  private parseThis(node: ThisExpression, vars: object) {
    if ({}.hasOwnProperty.call(vars, 'this')) {
      return (<any>vars)['this'];
    }
    return FAIL;
  }

  private parseCall(node: CallExpression, vars: object) {
    const callee = this.walk(node.callee, vars);
    if (callee === FAIL) {
      return FAIL;
    }
    if (typeof callee !== 'function') {
      return FAIL;
    }

    let ctx = (<any>node.callee).object
      ? this.walk((<any>node.callee).object, vars)
      : FAIL;
    if (ctx === FAIL) {
      ctx = null;
    }

    const args = [];
    for (const arg of node.arguments) {
      const x = this.walk(arg, vars);
      if (x === FAIL) {
        return FAIL;
      }
      args.push(x);
    }
    return callee.apply(ctx, args);
  }

  private parseMember(node: MemberExpression, vars: object) {
    const obj = this.walk(node.object, vars);
    if (obj === undefined) {
      return undefined;
    }

    if (obj === FAIL) {
      return FAIL;
    }
    if (node.property.type === 'Identifier') {
      return obj[node.property.name];
    }
    const prop = this.walk(node.property, vars);
    if (prop === FAIL) {
      return FAIL;
    }
    return obj[prop];
  }

  private parseConditional(node: ConditionalExpression, vars: object) {
    const val = this.walk(node.test, vars);
    if (val === FAIL) {
      return FAIL;
    }
    return val
      ? this.walk(node.consequent, vars)
      : this.walk(node.alternate, vars);
  }

  private parseStatement(node: ExpressionStatement, vars: object) {
    const val = this.walk(node.expression, vars);
    if (val === FAIL) {
      return FAIL;
    }
    return val;
  }

  private parseReturnStatement(node: ReturnStatement, vars: object) {
    return this.walk(node.argument, vars);
  }

  private parseFunction(node: FunctionExpression, vars: object) {
    const bodies = node.body.body;

    const newVars = {
      ...vars,
    };

    node.params.forEach((key) => {
      if (key.type === 'Identifier') {
        (<any>newVars)[key.name] = null;
      }
    });

    for (const i in bodies) {
      if (this.walk(bodies[i], newVars) === FAIL) {
        return FAIL;
      }
    }

    const keys = Object.keys(vars);
    const vals = keys.map((key) => {
      return (<any>vars)[key];
    });
    return Function(keys.join(', '), 'return ' + generate(node)).apply(
      null,
      vals,
    );
  }

  private parseTemplateLiteral(node: TemplateLiteral, vars: object) {
    let str = '';
    let i: number;
    for (i = 0; i < node.expressions.length; i++) {
      str += this.walk(node.quasis[i], vars);
      str += this.walk(node.expressions[i], vars);
    }
    str += this.walk(node.quasis[i], vars);
    return str;
  }

  private parseTaggedTemplate(node: TaggedTemplateExpression, vars: object) {
    const tag = this.walk(node.tag, vars);
    const quasi = node.quasi;
    const strings = quasi.quasis.map((q) => this.walk(q, vars));
    const values = quasi.expressions.map((e) => this.walk(e, vars));
    return tag.apply(null, [strings].concat(values));
  }

  private walk(
    node:
      | Expression
      | SpreadElement
      | Pattern
      | Super
      | TemplateElement
      | Statement
      | null
      | undefined,
    vars: object,
  ): any {
    if (!node) {
      return FAIL;
    }
    switch (node.type) {
      case 'Literal':
        return node.value;
      case 'UnaryExpression':
        return this.parseUnary(node, vars);
      case 'ArrayExpression':
        return this.parseArray(node, vars);
      case 'ObjectExpression':
        return this.parseObject(node, vars);
      case 'BinaryExpression':
        return this.parseBinary(node, vars);
      case 'LogicalExpression':
        return this.parseLogical(node, vars);
      case 'Identifier':
        return this.parseIdentifier(node, vars);
      case 'ThisExpression':
        return this.parseThis(node, vars);
      case 'CallExpression':
        return this.parseCall(node, vars);
      case 'MemberExpression':
        return this.parseMember(node, vars);
      case 'ConditionalExpression':
        return this.parseConditional(node, vars);
      case 'ExpressionStatement':
        return this.parseStatement(node, vars);
      case 'ReturnStatement':
        return this.parseReturnStatement(node, vars);
      case 'FunctionExpression':
        return this.parseFunction(node, vars);
      case 'TemplateLiteral':
        return this.parseTemplateLiteral(node, vars);
      case 'TaggedTemplateExpression':
        return this.parseTaggedTemplate(node, vars);
      case 'TemplateElement':
        return node.value.cooked;
      default:
        return FAIL;
    }
  }
}
