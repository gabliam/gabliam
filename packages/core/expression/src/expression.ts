import { Expression as AstExpression } from 'estree';
import { Parser, IS_STRING } from './parser';
import _ from 'lodash';
const validPath = require('is-valid-path');

export class Expression {
  private parser: Parser | undefined;

  constructor(
    ast: AstExpression | null,
    private context: object = {},
    private input: string
  ) {
    if (ast) {
      this.parser = new Parser(ast);
    }
  }

  getValue<T>(vars: object = {}): T | undefined | null {
    if (!this.parser) {
      if (validPath(this.input)) {
        return <any>this.input;
      }
      return undefined;
    }

    const context = Object.keys(vars).reduce(
      (prev, current) => {
        (<any>prev)[`$${current}`] = _.cloneDeep((<any>vars)[current]);
        return prev;
      },
      {
        $root: this.context,
      }
    );

    try {
      const res = this.parser.parse({
        ...this.context,
        ...context,
      });

      if (res === null && validPath(this.input)) {
        return <any>this.input;
      }

      if (res === IS_STRING) {
        return <any>this.input;
      }

      return res;
    } catch (e) {
      // case of multiple subpath
      if (validPath(this.input)) {
        return <any>this.input;
      }
      throw e;
    }
  }
}
