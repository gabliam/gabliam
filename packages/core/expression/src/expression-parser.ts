import * as esprima from 'esprima';
import { ExpressionStatement } from 'estree';
import _ from 'lodash';
import { Expression } from './expression';

export class ExpressionParser {
  static eval(input: string, vars: object = {}) {
    const expressionParser = new ExpressionParser({});
    return expressionParser.parseExpression(input).getValue(vars);
  }

  private context: object;

  constructor(context: object = {}) {
    this.context = _.cloneDeep(context);
  }

  parseExpression(input: string) {
    try {
      const program = esprima.parseScript(input);
      const ast = (<ExpressionStatement>program.body[0]).expression;
      return new Expression(ast, this.context, input);
    } catch {
      return new Expression(null, this.context, input);
    }
  }
}
