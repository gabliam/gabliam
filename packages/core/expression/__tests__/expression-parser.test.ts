import { ExpressionParser } from '../src';

describe('expression evaluator', () => {
  describe('compile', () => {
    test('should compile an expression an return an evaluator', () => {
      const compiledExpression = ExpressionParser.eval('1234');
      expect(compiledExpression).toMatchSnapshot();
    });
  });

  describe('parse', () => {
    describe('primitives', () => {
      test('should evaluate a number', () => {
        const numberInt = ExpressionParser.eval('123');
        const numberFloat = ExpressionParser.eval('123.4');

        expect(numberInt).toMatchSnapshot();
        expect(typeof numberInt).toMatchSnapshot();
        expect(numberFloat).toMatchSnapshot();
        expect(typeof numberFloat).toMatchSnapshot();
      });

      test('should evaluate a string', () => {
        // tslint:disable-next-line:quotemark
        const stringSingle = ExpressionParser.eval("'hello world!'");
        const stringDouble = ExpressionParser.eval('"hello world!"');

        expect(stringSingle).toMatchSnapshot();
        expect(typeof stringSingle).toMatchSnapshot();
        expect(stringDouble).toMatchSnapshot();
        expect(typeof stringSingle).toMatchSnapshot();
      });

      test('should evaluate a string with embedded escaped single quotes', () => {
        // tslint:disable-next-line:quotemark
        const stringSingle = ExpressionParser.eval("'hello world !'");

        const stringDouble = ExpressionParser.eval('"hello world !"');

        expect(stringSingle).toMatchSnapshot();
        expect(typeof stringSingle).toMatchSnapshot();
        expect(stringDouble).toMatchSnapshot();
        expect(typeof stringDouble).toMatchSnapshot();
      });

      test('should evaluate a string with embedded escaped double quotes', () => {
        // tslint:disable-next-line:quotemark
        const stringSingle = ExpressionParser.eval("'hello world!'");
        const stringDouble = ExpressionParser.eval('"hello world!"');

        expect(stringSingle).toMatchSnapshot();
        expect(typeof stringSingle).toMatchSnapshot();
        expect(stringDouble).toMatchSnapshot();
        expect(typeof stringDouble).toMatchSnapshot();
      });

      test('should evaluate a boolean', () => {
        const boolTrue = ExpressionParser.eval('true');
        const boolFalse = ExpressionParser.eval('false');
        expect(boolTrue).toMatchSnapshot();
        expect(typeof boolTrue).toMatchSnapshot();
        expect(boolFalse).toMatchSnapshot();
        expect(typeof boolFalse).toMatchSnapshot();
      });
    });
  });

  describe('lookups', () => {
    let expressionParser: ExpressionParser;

    beforeAll(() => {
      expressionParser = new ExpressionParser({
        iAmANumber: 1,
        iAmANestedPropertyName: 'propLookup',
        nested: {
          iAmAString: 'hi',
          reallyNested: {
            iAmTrue: true,
            hi: 'bye',
          },
          propLookup: 'Found!',
        },
      });
    });

    test('should look up a primitive in the context', () => {
      const n = expressionParser.parseExpression('iAmANumber').getValue();
      expect(n).toMatchSnapshot();
      expect(typeof n).toMatchSnapshot();
    });

    test('should look up a nested primitive in the context using dot notation', () => {
      const s = expressionParser
        .parseExpression('nested.iAmAString')
        .getValue();

      expect(s).toMatchSnapshot();
      expect(typeof s).toMatchSnapshot();
    });

    test('should look up a doubly nested primitive in the context using dot notation', () => {
      const bool = expressionParser
        .parseExpression('nested.reallyNested.iAmTrue')
        .getValue();

      expect(bool).toMatchSnapshot();
      expect(typeof bool).toMatchSnapshot();
    });

    test('should look up a nested primitive in the context using bracket notation literal', () => {
      const s = expressionParser
        .parseExpression('nested["iAmAString"]')
        .getValue();
      expect(s).toMatchSnapshot();
      expect(typeof s).toMatchSnapshot();
    });

    test('should look up a nested primitive in the context using bracket notation', () => {
      const s = expressionParser
        .parseExpression('nested[iAmANestedPropertyName]')
        .getValue();
      expect(s).toMatchSnapshot();
      expect(typeof s).toMatchSnapshot();
    });

    test('should look up a really nested primitive in the context using bracket notation', () => {
      const s = expressionParser
        .parseExpression('nested.reallyNested[nested.iAmAString]')
        .getValue();
      expect(s).toMatchSnapshot();
      expect(typeof s).toMatchSnapshot();
    });

    test('should return undefined', () => {
      const s = expressionParser
        .parseExpression('nested.doestExist')
        .getValue();
      expect(s).toMatchSnapshot();
      expect(typeof s).toMatchSnapshot();
    });
  });

  describe('comparisons', () => {
    test('should evaluate an equality', () => {
      const comp1 = ExpressionParser.eval('1 == 1');
      const comp2 = ExpressionParser.eval('1 == 2');
      const comp3 = ExpressionParser.eval(`1 == '1'`);

      expect(comp1).toBe(true);
      expect(comp2).toBe(false);
      expect(comp3).toBe(true);
    });

    test('should evaluate an equality with lookups', () => {
      const context = {
        left: 1,
        right: 1,
      };

      const comp = ExpressionParser.eval('$left == $right', context);

      expect(comp).toBe(true);
    });

    test('should evaluate an inequality (not equal)', () => {
      const comp1 = ExpressionParser.eval('1 != 2');
      const comp2 = ExpressionParser.eval('1 != 1');
      const comp3 = ExpressionParser.eval(`1 != '1'`);

      expect(comp1).toBe(true);
      expect(comp2).toBe(false);
      expect(comp3).toBe(false);
    });

    test('should evaluate an equality ===', () => {
      const comp1 = ExpressionParser.eval('1 === 1');
      const comp2 = ExpressionParser.eval('1 === 2');
      const comp3 = ExpressionParser.eval(`1 === '1'`);

      expect(comp1).toBe(true);
      expect(comp2).toBe(false);
      expect(comp3).toBe(false);
    });

    test('should evaluate an inequality (not equal) !==', () => {
      const comp1 = ExpressionParser.eval('1 !== 2');
      const comp2 = ExpressionParser.eval('1 !== 1');
      const comp3 = ExpressionParser.eval(`1 !== '1'`);

      expect(comp1).toBe(true);
      expect(comp2).toBe(false);
      expect(comp3).toBe(true);
    });

    test('should evaluate an inequality (greater than)', () => {
      const comp1 = ExpressionParser.eval('2 > 1');
      const comp2 = ExpressionParser.eval('1 > 1');

      expect(comp1).toBe(true);
      expect(comp2).toBe(false);
    });

    test('should evaluate an inequality (greater than or equal to)', () => {
      const comp1 = ExpressionParser.eval('1 >= 1');
      const comp2 = ExpressionParser.eval('2 >= 1');
      const comp3 = ExpressionParser.eval('1 >= 2');

      expect(comp1).toBe(true);
      expect(comp2).toBe(true);
      expect(comp3).toBe(false);
    });

    test('should evaluate an inequality (less than)', () => {
      const comp1 = ExpressionParser.eval('1 < 2');
      const comp2 = ExpressionParser.eval('1 < 1');

      expect(comp1).toBe(true);
      expect(comp2).toBe(false);
    });

    test('should evaluate an inequality (less than or equal to)', () => {
      const comp1 = ExpressionParser.eval('1 <= 2');
      const comp2 = ExpressionParser.eval('1 <= 2');
      const comp3 = ExpressionParser.eval('2 <= 1');

      expect(comp1).toBe(true);
      expect(comp2).toBe(true);
      expect(comp3).toBe(false);
    });

    test('should evaluate a complex inequality', () => {
      const comp = ExpressionParser.eval('"abc".length <= "abcde".length');

      expect(comp).toBe(true);
    });
  });

  describe('unary', () => {
    test('+', () => {
      expect(ExpressionParser.eval('+3')).toBe(3);
      expect(ExpressionParser.eval('+"3"')).toBe(3);
      expect(ExpressionParser.eval('+true')).toBe(1);
      expect(ExpressionParser.eval('+false')).toBe(0);
      expect(ExpressionParser.eval('+null')).toBe(0);
      expect(ExpressionParser.eval('+function(val){ return val; }')).toBeNaN();
    });

    test('+ with context', () => {
      const expression = new ExpressionParser({
        a: 3,
        b: '3',
        c: true,
        d: false,
        e: null,
        f: function(val: any) {
          return val;
        },
      });
      expect(expression.parseExpression('+a').getValue()).toBe(3);
      expect(expression.parseExpression('+b').getValue()).toBe(3);
      expect(expression.parseExpression('+c').getValue()).toBe(1);
      expect(expression.parseExpression('+d').getValue()).toBe(0);
      expect(expression.parseExpression('+e').getValue()).toBe(0);
      expect(expression.parseExpression('+f').getValue()).toBeNaN();
    });

    test('-', () => {
      expect(ExpressionParser.eval('-(3)')).toBe(-3);
      expect(ExpressionParser.eval('-(-3)')).toBe(3);
      expect(ExpressionParser.eval('-"4"')).toBe(-4);
    });

    test('- with context', () => {
      const expression = new ExpressionParser({
        a: 3,
        b: -3,
        c: '4',
      });
      expect(expression.parseExpression('-a').getValue()).toBe(-3);
      expect(expression.parseExpression('-b').getValue()).toBe(3);
      expect(expression.parseExpression('-c').getValue()).toBe(-4);
    });
  });

  describe('method invocation', () => {
    let expressionParser: ExpressionParser;
    beforeAll(() => {
      expressionParser = new ExpressionParser({
        funky: () => {
          return 'fresh';
        },
        argumentative: (arg: any) => {
          return arg;
        },
        name: 'ben',
      });
    });

    test('should look up and invoke a function', () => {
      const ret = expressionParser.parseExpression('funky()').getValue();

      expect(ret).toBe('fresh');
    });

    test('should look up and invoke a function with arguments', () => {
      const ret = expressionParser
        .parseExpression('argumentative("i disagree!")')
        .getValue();

      expect(ret).toBe('i disagree!');
    });

    test('should use a property if getter not available', () => {
      const ret = expressionParser.parseExpression('name').getValue();

      expect(ret).toBe('ben');
    });
  });

  describe('logical', () => {
    it('should return false and no throw error', () => {
      const local = new ExpressionParser()
        .parseExpression('Array.isArray($result) && $result.length !== 0')
        .getValue({
          result: null,
        });

      expect(local).toBe(false);
    });
  });

  describe('locals', () => {
    it('should refer to a local variable', () => {
      const local = new ExpressionParser({
        myString: 'global context',
      })
        .parseExpression('$myString == "hello world!"')
        .getValue({
          myString: 'hello world!',
        });

      expect(local).toBe(true);
    });

    it('should refer to the root context', () => {
      const local = new ExpressionParser({
        myString: 'global context',
      })
        .parseExpression('$root')
        .getValue({
          myString: 'hello world!',
        });

      expect(local).toMatchSnapshot();
    });
  });

  describe('math', () => {
    it('should add 2 numbers', () => {
      const sum = ExpressionParser.eval('1 + 1');

      expect(sum).toBe(2);
    });

    it('should add 3 numbers', () => {
      const sum = ExpressionParser.eval('1 + 1 + 1');

      expect(sum).toBe(3);
    });

    it('should subtract 2 numbers', () => {
      const difference = ExpressionParser.eval('1 - 1');

      expect(difference).toBe(0);
    });

    it('should multiply 2 numbers', () => {
      const product = ExpressionParser.eval('2 * 2');

      expect(product).toBe(4);
    });

    it('should divide 2 numbers', () => {
      const quotient = ExpressionParser.eval('4 / 2');

      expect(quotient).toBe(2);
    });

    it('should find the modulus of 2 numbers', () => {
      const mod = ExpressionParser.eval('10 % 8');

      expect(mod).toBe(2);
    });

    it('should evaluate an exponent', () => {
      const mod = ExpressionParser.eval('10^2');
      expect(mod).toBe(8);
    });

    it('should honor standard order of operations', () => {
      const math = ExpressionParser.eval('8 + 4 * 6 - 2 * 3 / 2'); // 8+(4*6)-(2*3/2) = 29

      expect(math).toBe(29);
    });
  });

  describe('ternary', () => {
    it('should return first argument if true', () => {
      const tern = ExpressionParser.eval('true ? "yes" : "no"');

      expect(tern).toBe('yes');
    });

    it('should return second argument if false', () => {
      const tern = ExpressionParser.eval('false ? "yes" : "no"');

      expect(tern).toBe('no');
    });
  });

  describe('complex literals', () => {
    it('should create an array', () => {
      const arr = ExpressionParser.eval('[1, 2, 3, 4]');
      expect(arr).toEqual([1, 2, 3, 4]);
    });

    // it.only('should create a map', () => {
    //   const map = ExpressionParser.eval('val = { name: "Nikola", dob: "10-July-1856" }');

    //   expect(map).toEqual({ name: 'Nikola', dob: '10-July-1856' });
    // });
  });

  describe('path', () => {
    it('with no subpath', () => {
      const str = ExpressionParser.eval('/photo');
      expect(str).toEqual('/photo');
    });

    it('with no subpath and end slash', () => {
      const str = ExpressionParser.eval('/photo/');
      expect(str).toEqual(/photo/);
    });

    it('with 1 subpath', () => {
      const str = ExpressionParser.eval('/photo/lol');
      expect(str).toEqual('/photo/lol');
    });

    it('with 1 subpath and end slash', () => {
      const str = ExpressionParser.eval('/photo/lol/');
      expect(str).toEqual('/photo/lol/');
    });

    it('with 2 subpath', () => {
      const str = ExpressionParser.eval('/photo/lol/lol');
      expect(str).toEqual('/photo/lol/lol');
    });

    it('with 2 subpath and end slash', () => {
      const str = ExpressionParser.eval('/photo/lol/lol/');
      expect(str).toEqual('/photo/lol/lol/');
    });

    it('with 3 subpath', () => {
      const str = ExpressionParser.eval('/photo/lol/lol/lol');
      expect(str).toEqual('/photo/lol/lol/lol');
    });

    it('with 3 subpath and end slash', () => {
      const str = ExpressionParser.eval('/photo/lol/lol/lol/');
      expect(str).toEqual('/photo/lol/lol/lol/');
    });
  });
});
