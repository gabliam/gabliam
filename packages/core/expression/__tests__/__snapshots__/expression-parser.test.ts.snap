// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expression evaluator compile should compile an expression an return an evaluator 1`] = `1234`;

exports[`expression evaluator locals should refer to the root context 1`] = `
{
  "myString": "global context",
}
`;

exports[`expression evaluator lookups should look up a doubly nested primitive in the context using dot notation 1`] = `true`;

exports[`expression evaluator lookups should look up a doubly nested primitive in the context using dot notation 2`] = `"boolean"`;

exports[`expression evaluator lookups should look up a nested primitive in the context using bracket notation 1`] = `undefined`;

exports[`expression evaluator lookups should look up a nested primitive in the context using bracket notation 2`] = `"undefined"`;

exports[`expression evaluator lookups should look up a nested primitive in the context using bracket notation literal 1`] = `"hi"`;

exports[`expression evaluator lookups should look up a nested primitive in the context using bracket notation literal 2`] = `"string"`;

exports[`expression evaluator lookups should look up a nested primitive in the context using dot notation 1`] = `"hi"`;

exports[`expression evaluator lookups should look up a nested primitive in the context using dot notation 2`] = `"string"`;

exports[`expression evaluator lookups should look up a primitive in the context 1`] = `1`;

exports[`expression evaluator lookups should look up a primitive in the context 2`] = `"number"`;

exports[`expression evaluator lookups should look up a really nested primitive in the context using bracket notation 1`] = `"bye"`;

exports[`expression evaluator lookups should look up a really nested primitive in the context using bracket notation 2`] = `"string"`;

exports[`expression evaluator lookups should return undefined 1`] = `undefined`;

exports[`expression evaluator lookups should return undefined 2`] = `"undefined"`;

exports[`expression evaluator parse primitives should evaluate a boolean 1`] = `true`;

exports[`expression evaluator parse primitives should evaluate a boolean 2`] = `"boolean"`;

exports[`expression evaluator parse primitives should evaluate a boolean 3`] = `false`;

exports[`expression evaluator parse primitives should evaluate a boolean 4`] = `"boolean"`;

exports[`expression evaluator parse primitives should evaluate a number 1`] = `123`;

exports[`expression evaluator parse primitives should evaluate a number 2`] = `"number"`;

exports[`expression evaluator parse primitives should evaluate a number 3`] = `123.4`;

exports[`expression evaluator parse primitives should evaluate a number 4`] = `"number"`;

exports[`expression evaluator parse primitives should evaluate a string 1`] = `"hello world!"`;

exports[`expression evaluator parse primitives should evaluate a string 2`] = `"string"`;

exports[`expression evaluator parse primitives should evaluate a string 3`] = `"hello world!"`;

exports[`expression evaluator parse primitives should evaluate a string 4`] = `"string"`;

exports[`expression evaluator parse primitives should evaluate a string with embedded escaped double quotes 1`] = `"hello world!"`;

exports[`expression evaluator parse primitives should evaluate a string with embedded escaped double quotes 2`] = `"string"`;

exports[`expression evaluator parse primitives should evaluate a string with embedded escaped double quotes 3`] = `"hello world!"`;

exports[`expression evaluator parse primitives should evaluate a string with embedded escaped double quotes 4`] = `"string"`;

exports[`expression evaluator parse primitives should evaluate a string with embedded escaped single quotes 1`] = `"hello world !"`;

exports[`expression evaluator parse primitives should evaluate a string with embedded escaped single quotes 2`] = `"string"`;

exports[`expression evaluator parse primitives should evaluate a string with embedded escaped single quotes 3`] = `"hello world !"`;

exports[`expression evaluator parse primitives should evaluate a string with embedded escaped single quotes 4`] = `"string"`;
